syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+            item.pop(key, None)
+        ts = item.get("modified")
+        if ts is not None:
+            item["modified"] = dt2ts(ts)
+        # Convert the ttl back into an offset from the current time.
+        ttl = item.get("ttl")
+        if ttl is not None:
+            item["ttl"] = int(dt2ts(ttl) - timestamp)
+        return BSO(item)
+
+    @with_session
+    def set_items(self, session, user, collection, items):
+        """Creates or updates multiple items in a collection."""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection, create=True)
+        self._touch_collection(session, userid, collectionid)
+        if not items:
+            return session.timestamp
+
+        item_ids = [x["id"] for x in items]
+        bind_names = ["@id_%d" % x for x in range(1, len(items) + 1)]
+        bind = {"userid": userid,
+                "collectionid": collectionid}
+        bind_types = {"userid": param_types.STRING,
+                      "collectionid": param_types.INT64}
+        for i, id in enumerate(item_ids, 1):
+            bind["id_%d" % (i,)] = id
+            bind_types["id_%d" % (i,)] = param_types.STRING
+        q = """\
+        SELECT id FROM bso
+        WHERE userid=@userid AND collection=@collectionid AND id in (%s)"""
+        q = q % ', '.join(bind_names)
+        result = session.transaction.execute_sql(
+            q,
+            params=bind,
+            param_types=bind_types
+        )
+        existing = {row[0] for row in result}
+
+        insert_rows = []
+        updates = []
+        for item in items:
+            if item["id"] in existing:
+                updates.append(item)
+                continue
+            row = self._prepare_bso_row(
+                session,
+                userid,
+                collectionid,
+                item["id"],
+                item
+            )
+            insert_rows.append(row)
+
+        if insert_rows:
+            session.transaction.insert(
+                "bso",
+                columns=["userid", "collection", "id", "sortindex", "modified",
+                         "payload", "ttl"],
+                values=insert_rows)
+
+        for item in updates:
+            cols, vals = self._prepare_update_bso_row(
+                session,
+                userid,
+                collectionid,
+                item["id"],
+                item
+            )
+            session.transaction.update(
+                "bso",
+                columns=cols,
+                values=[vals]
+            )
+        return session.timestamp
+
+    @with_session
+    def create_batch(self, session, user, collection):
+        """Creates a batch in batch_uploads table"""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        # TODO: raise ConflictError if the insert fails for an existing
+        # TODO: add the batchid skew (user % 10) or use the full microsecond TS
```

Comment:
Lol it'd probably be quite nice if you didn't have to repeat that hackery for this batch implementation 😆